Noise Plethora (v2.1.0)

hemmer opened the pull request:
PR for #29

hemmer committed:
Move prototype Noise Plethora out of private repo

hemmer committed:
Implement non-arm based multiply, and try out on CrossModRing

hemmer committed:
Major refactor to make all plugins use block-based processing

* remove some debug outputs
* fixed Atari plugin not working
* removed unused float based replacements
* formatting

hemmer committed:
Cleaning up some compiler warnings

Trying to address some possible Mac/Linux bugs

Splitting more effects into cpp/hpp

hemmer committed:
Github workflow fix

hemmer committed:
Fix runtime error on Mac/Linux (functions not marked pure virtual)

hemmer committed:
Fix off by one error with bank selection

Remove test Bank

Some unused variable warnings

hemmer committed:
Noise Plethora: big refactor for readability

* added DC blocking (4th order Butterworth HP at 10Hz)
* updated Chopping Kinky to use this
* removed debug messages
* bump SDK version
* addressing some compiler warning
* improved filter efficiency

hemmer committed:
Added soft clipping to outputs

* added asymmetric clipping to Section C out
* tuning filters
* adding DC block to Section C
* tooltips for inputs/outputs etc

hemmer committed:
Fix playability of resonances and grit quantity

hemmer committed:
Tweak DC blocker, reinitialise algorithms onSampleRateChange()

hemmer committed:
Added technical description of Teensy replacements, and checks for JSON reading

hemmer committed:
Added bypass for Spring Reverb, and minor bugfix for Kickall triggers

hemmer committed:
Update license info for Noise Plethora plugins

hemmer committed:
Pin down max frequencies so Noise Plethora is more robust to sample rate changes

hemmer commented on LICENSE.md:
```diff
@@ -1,6 +1,8 @@
 All **source code** is copyright © 2019-2021 [VCV](https://vcvrack.com/) and contributers.
 
-This program is free software: you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-3.0.en.html) as published by the [Free Software Foundation](https://www.fsf.org/), either version 3 of the License, or (at your option) any later version.
+This program is free software: you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-3.0.en.html) as published by the [Free Software Foundation](https://www.fsf.org/), version 3 of the License (GPL-3.0-only).
```
change to `GPL-3.0-only` to be compatible with Noise Plethora CC-BY-SA v4 license used for plugins.

falkTX replied:
very curious here, how is gplv3+ incompatible with cc-by-sa?

hemmer replied:
See: https://creativecommons.org/share-your-work/licensing-considerations/compatible-licenses

Also last paragraph of https://wiki.creativecommons.org/wiki/ShareAlike_compatibility:_GPLv3

Basically, GPL version can be increased at a later date if determined compatible with CC, but for now GPLv3 only.

If I've misunderstood then let me know!

falkTX replied:
That doesnt invalidate the code "or later" clause as I see it.
And that whole compatible license thing is for when generating modifications of the artwork files, say someone mixing *artwork* that is CC-BY with *artwork* that is GPL.

So as I see it, the usage within Rack modules does not apply, as the separation between artwork and code is quite clear.
Otherwise one would not be able to use CC-NC or CC-ND like Rack itself or Fundamental does, as those licenses are inherently incompatible with GPL.

hemmer replied:
Just to clarify, I am adapting code (the code for each of the 30 Noise Plethora plugin) not artwork in this instance. The relevant lines (emphasis mine) are:

> You must specify version 3 of the GPL when redistributing a GPL project into which BY-SA 4.0 content is adapted.

> Because this compatibility determination is **limited to version 3 of the GPL**, reusers that adapt BY-SA 4.0 content and incorporate it into GPL-licensed projects are only able to use GPLv3 unless and until other GPL versions are declared compatible.

falkTX replied:
oh so the actual code this is based on is CC licensed? that is very odd choice. but then would make sense yes

hemmer replied:
exactly:

license text (last sentence): https://github.com/Befaco/Noise_plethora/blob/master/README.md
example plugin: https://github.com/Befaco/Noise_plethora/blob/master/P_TriFMcluster.hpp

dromer replied:
But this license change would be only for this specific module, then?
Doesn't make a lot of sense to re-license the entire project like this.

falkTX replied:
all the code is merged together at the end as single binary, so the most restrictive of licenses will apply.
if there is gplv3+ and gplv3-only on the same project, final binary is gplv3-only

dromer replied:
True-true, wasn't paying attention. Still not sure if it makes sense to merge mixed licensed stuff like this and in essence "downgrade" the existing code.

hemmer replied:
Team Befaco have [relicensed](https://github.com/Befaco/Noise_Plethora) the plugin files to GPL-v3-or-later, so we can go back to GPL-v3-or-later. This PR should now reflect the updated license.

hemmer commented on src/ChoppingKinky.cpp:
```diff
@@ -50,7 +50,7 @@ struct ChoppingKinky : Module {
 	int oversamplingIndex = 2; 	// default is 2^oversamplingIndex == x4 oversampling
 
-	dsp::BiquadFilter blockDCFilter;
+	DCBlocker blockDCFilter;
```
Replaced with generic DC blocker (as DC blocking in a few places now).

hemmer commented on src/NoisePlethora.cpp:
```diff
@@ -0,0 +1,892 @@
+		maxAngle = 0.80 * M_PI;
+	}
+
+	// this seems convoluted but I can't see how to achieve the following (say) with onAction:
```
If there's an easier way, let me know, but I couldn't see how to get all three...

hemmer commented on src/SpringReverb.cpp:
```diff
@@ -71,6 +71,16 @@ struct SpringReverb : Module {
 		delete convolver;
 	}
 
+	void processBypass(const ProcessArgs& args) override {
```
user request for bypass on reverb

hemmer committed:
Remove github yml, update version, tooltips for A*B+C

hemmer commented on src/noise-plethora/README.md:
```diff
@@ -0,0 +1,17 @@
+Technical Notes
```
just in case you are interested...

hemmer commented:
Hi @AndrewBelt - I'm sure you're super busy with V2 stuff, but if you have a moment to have a look over some of the key changes (e.g. license changes) / checkout the code and have a play then great. There are too many files for any meaningful review, but you might be interested in say the core Noise Plethora plugin. If you are too busy, but _are_ happy for me to merge the changes in regardless then do say.

hemmer commented on src/noise-plethora/LICENSE.md:
```diff
@@ -0,0 +1,2 @@
+Plugins in `./plugins` have been derived from source files from https://github.com/Befaco/Noise_plethora/ (prefix `P_*.hpp`) which were originally licensed under CC-BY-SA v4 (see https://github.com/Befaco/Noise_plethora/blob/master/README.md). This license is [one-way compatible with GPL-v3-only](https://creativecommons.org/share-your-work/licensing-considerations/compatible-licenses), which the  ...
+
```
I should probably add that the modified Teensy codes are MIT licensed.

hemmer replied:
Added

AndrewBelt replied:
Are Befaco and Teensy licenses included in the binary package? If not, add them to `LICENSE-dist.txt` or something and add that file to `DISTRIBUTABLES += ...`

hemmer replied:
Will do. Befaco Noise Plethora sources are now GPL-v3-or-later (same as this repo), does that still needed to be added?

hemmer committed:
Added note about Teensy audio components license

hemmer committed:
Fix popping on increase of polyphony channels (EvenVCO)

Fixes #33

hemmer committed:
Fix crash in Ableton (font reference across multiple frames)

hemmer commented:
Fixed crash in VST version (font references across multiple frames), fix added to PR.

hemmer committed:
Fix Muxlicer bug when getting repeated first step, added Hexmix VCA tooltips/bypasses.

Fixes #32. Couple of things going on - one is that the 1ms grace period on reset doesn't really work in this context, and breaks EOC chaining which is a common use case. Two, if the clock is already high when a reset was requested, then the boolean trigger detector doesn't detect a rising edge.

hemmer committed:
More more on fixing Muxlicer reset bug

hemmer committed:
Update license to GPLv3 or later

Team Befaco have upgraded the license on the source Noise_Plethora repo so we can go back to using GPL-v3-or-later

AndrewBelt commented on src/EvenVCO.cpp:
```diff
@@ -120,8 +120,10 @@ struct EvenVCO : Module {
-		for (int c = 0; c < channels; c++) {
+		// the next block can't be done with SIMD instructions, but should at least be completed with
+		// blocks of 4 (otherwise popping artfifacts are generated from invalid phase/oldPhase/deltaPhase)
+		int channelsRoundedUpNearestFour = std::ceil(channels / 4.f) * 4;
```
Why convert to float and back to int when you can just use int arithmetic?

AndrewBelt commented on CHANGELOG.md:
```diff
@@ -1,9 +1,18 @@
+  * Added bypass for Spring Reverb
+  * Bugfix: Kickall trigger input and button can now work independently
+  * Bugfix: EvenVCO remove pop on number of polyphony engines change
+  * Bugfix: Muxlicer chaining using reset now works correctly
```
Can you group changes into modules?

AndrewBelt replied:
No need to use the word "bugfix", use the language "Fix XYZ trigger input" etc.

AndrewBelt commented on src/ABC.cpp:
```diff
@@ -45,6 +45,17 @@ struct ABC : Module {
 		configParam(C2_LEVEL_PARAM, -1.0, 1.0, 0.0, "C2 Level");
+
+		configInput(A1_INPUT, "A1");
+		configInput(B1_INPUT, "B1 (normalled to 5V)");
```
Remove the description from ports and add it to
```cpp
paramQuantities[XYZ_INPUT]->description = "Normalled to 5V";
```

hemmer replied:
I guess in this case the inputs have a corresponding param (knob), but that's not always the case. Unless I'm missing something, it doesn't look like inputs can have an independent description from their name. It's a shame to lose this information (as I often find myself wondering what CV range a module will accept), but happy to go with your preference.

If you meant update the corresponding knob to say, e.g., "normalled to 10V", I don't think that makes any more sense.

AndrewBelt replied:
My apologies, my code should be
```cpp
getInputInfo(B1_INPUT)->description = "Normalled to 5V";
```
Port tooltips should show descriptions too. Let me know if it doesn't work.

hemmer replied:
Excellent that works as expected, nice feature (especially for new users)!

AndrewBelt commented on src/HexmixVCA.cpp:
```diff
@@ -42,8 +42,14 @@ struct HexmixVCA : Module {
+			configParam(VOL_PARAM + i, 0.f, 1.f, 1.f, string::f("Channel %d output level", i + 1));
+
+			configInput(IN_INPUT + i, string::f("Channel %d", i + 1));
+			configInput(CV_INPUT + i, string::f("Gain %d (10V normalled)", i + 1));
```
Don't put descriptions in port names.

hemmer committed:
Addressing review feedback for pull request

Fix slightly janky behaviour of program knob for Noise Plethora

hemmer committed:
Make Nose Plethora algorithms addressable by CV in a consistent way

Add input port descriptions

hemmer committed:
Further improvements to CV control

hemmer committed:
Reverse CV LED colours on Noise Plethora
